lib/view_component/slotable.rb

```diff
@@ -395,6 +397,31 @@ def set_slot(slot_name, slot_definition = nil, *args, &block)
 
       slot
     end
+
+    def add_slot_iterator_options(options, component, slot_name)
+      if component.counter_argument_present? || component.iteration_argument_present?
+        @__vc_set_slots ||= {}
+        @collection_iterator ||=
+          ActionView::PartialIteration.new(
+            Array(
+              Hash(@__vc_set_slots)[slot_name]
+            ).length
+          )
+
+        if component.counter_argument_present?
+          options[-1][component.collection_counter_parameter] = @collection_iterator.index
+        end
+
+        if component.iteration_argument_present?
+          options[-1][component.collection_iteration_parameter] = @collection_iterator.dup
+        end
+
+        @collection_iterator.iterate!
+      end
+
+      options
+    end
```

Comment:
I like the changes that have been made to this method!

I do think it's a litle confusing, though. We're mutating the options argument on lines 412 and 416 and then returning the mutated arguments. Usually if the method mutates the objects you're passing in it doesn't then return them since you still have the original mutated references. And even if it did mutate and return, it would be for chaining convenience and not for how we're using it at the call site.

So the method use is telling me that it's not mutating the arguments but the body does mutate the arguments, which is a little mixed up, I think.

I think it would be better to avoid mutating the objects on 412 and 416 by adding on 410:

```ruby
options = options.dup.tap{|o| o[-1] = o[-1].dup }
```

I can go into a few reasons I think it's worth avoiding mutation if you'd like, but it's probably good enough to say that if the user passes in frozen objects this will raise an error.

Reply:
@corasaurus-hex I agree that their is a danger of mutating an array that is passed to you because it may be frozen, so I tried to implement the duped array aswell as the nested hash but the nested hash when duped does not work with the splat operator for soem reason, here teh error I get when I splat the hash
```
ArgumentError: wrong number of arguments (given 1, expected 0; required keywords: title, numbered_tab_counter, numbered_tab_iteration)

Reply:
on another note the args is not passed as an array to this method are not passed by the user bu rather the individual args of the with_slot_name accesor so it I do not think it is reasonable to expect the args to be in a immutable state from being frozen, etc

Reply:
How about this?

```diff
diff --git a/lib/view_component/slotable.rb b/lib/view_component/slotable.rb
index 8aa36a8b..ec637128 100644
--- a/lib/view_component/slotable.rb
+++ b/lib/view_component/slotable.rb
@@ -357,13 +357,13 @@ module ViewComponent

       # If class
       if slot_definition[:renderable]
-        args = add_slot_iterator_args(args, slot_definition[:renderable], slot_name)
-        slot.__vc_component_instance = slot_definition[:renderable].new(*args)
+        *args, opts = add_slot_iterator_options(args, slot_definition[:renderable], slot_name)
+        slot.__vc_component_instance = slot_definition[:renderable].new(*args, **opts)
       # If class name as a string
       elsif slot_definition[:renderable_class_name]
-        args = add_slot_iterator_args(args, self.class.const_get(slot_definition[:renderable_class_name]), slot_name)
+        *args, opts = add_slot_iterator_options(args, self.class.const_get(slot_definition[:renderable_class_name]), slot_name)
         slot.__vc_component_instance =
-          self.class.const_get(slot_definition[:renderable_class_name]).new(*args)
+          self.class.const_get(slot_definition[:renderable_class_name]).new(*args, **opts)
       # If passed a lambda
       elsif slot_definition[:renderable_function]
         # Use `bind(self)` to ensure lambda is executed in the context of the
@@ -399,9 +399,15 @@ module ViewComponent
       slot
     end

-    def add_slot_iterator_args(args, component, slot_name)
-      args = args.deep_dup
-      return args unless args.any?
+    def add_slot_iterator_options(args, component, slot_name)
+      args = args.dup
+
+      if args.last.is_a?(Hash)
+        args[-1] = args[-1].dup
+      else
+        args << {}
+      end
+
       if component.counter_argument_present? || component.iteration_argument_present?
         @__vc_set_slots ||= {}
         @collection_iterator ||=
```